Python/fileutils.c

```diff
@@ -922,6 +922,86 @@ _Py_GetLocaleEncodingObject(void)
     return str;
 }
 
+#ifdef HAVE_NON_UNICODE_WCHAR_T_REPRESENTATION
+
+/* Convert a wide character string to the UTF32 encoded char32_t string. This
+   is necessary on systems where internal form of wchar_t is not already
+   Unicode (e.g. Oracle Solaris).
+
+   Return a pointer to a newly allocated char32_t string, use PyMem_Free() to
+   free the memory. Return NULL and raise exception on conversion or memory
+   allocation error. */
+char32_t*
+_Py_convert_wchar_t_to_UTF32(const wchar_t* u, Py_ssize_t size)
+{
+    /* Ensure we won't overflow the size. */
+    if (size > ((PY_SSIZE_T_MAX / (Py_ssize_t)sizeof(wchar_t)) - 1)) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    /* Given 'u' might not be NULL terminated (size smaller than its
+       length); copy and terminate part we are interested in. */
+    wchar_t* substr = PyMem_Malloc((size + 1) * sizeof(wchar_t));
+    if (substr == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    memcpy(substr, u, size * sizeof(wchar_t));
+    substr[size] = 0;
+
+    /* Convert given wide-character string to a character string */
+    size_t buffsize = wcstombs(NULL, substr, 0) + 1;
```

Comment:
test_unicode.test_aswidechar() ensures that PyUnicode_FromWideChar() does not truncate the input string at null character. But wcstombs() stops at the first null character, no?

Reply:
Yes, it does - I was not aware that null characters should not truncate the input string, and the test you mentioned did not catch this because it runs under utf-8 locale (where my changes have no effect). I will look into it.

Reply:
I suggest you to use the same code to convert the string (loop using mbrtoc32 below) and to compute the buffer size. For example, write an helper function which is first called to compute the buffer size (NULL buffer), and then to actually convert the string (non-NULL buffer).

The loop below doesn't truncate at the first NULL characters and so can be reused.

Reply:
Solved with `iconv` which doesn't truncate.